### PR TITLE
Fix not escaping special characters in search pattern

### LINKED ISSUE #5892: Searching for a DOI results in an ERROR
JabRef 5.0-beta.388--2020-01-29--eb988b6
Windows 7 6.1 amd64 
Java 13.0.2
- [+] I have tested the latest development version from http://builds.jabref.org/master/ and the problem persists

I searched in an empty library (in my standard library as well) for the DOI 10.1210/endrev/bnz006. This resulted in an error. In my opinion this a serious failure since a DOI should be searchable. 

<details>
  <summary>Log File</summary>
  
  ```
ERROR FallbackExceptionHandler Uncaught exception occurred in Thread[JavaFX Application Thread,5,main]
netscape.javascript.JSException: SyntaxError: Invalid regular expression: missing )
       at org.jabref.merged.module/com.sun.webkit.dom.JSObject.fwkMakeException(Unknown Source)
       at org.jabref.merged.module/com.sun.webkit.WebPage.twkExecuteScript(Native Method)
       at org.jabref.merged.module/com.sun.webkit.WebPage.executeScript(Unknown Source)
       at org.jabref.merged.module/javafx.scene.web.WebEngine.executeScript(Unknown Source)
       at org.jabref/org.jabref.gui.preview.PreviewViewer.highlightSearchPattern(Unknown Source)
       at org.jabref/org.jabref.gui.preview.PreviewViewer.lambda$new$0(Unknown Source)
       at org.jabref.merged.module/com.sun.javafx.binding.ExpressionHelper$Generic.fireValueChangedEvent(Unknown Source)
       at org.jabref.merged.module/com.sun.javafx.binding.ExpressionHelper.fireValueChangedEvent(Unknown Source)
       at org.jabref.merged.module/javafx.beans.property.ObjectPropertyBase.fireValueChangedEvent(Unknown Source)
       at org.jabref.merged.module/javafx.beans.property.ObjectPropertyBase.markInvalid(Unknown Source)
       at org.jabref.merged.module/javafx.beans.property.ObjectPropertyBase.set(Unknown Source)
       at org.jabref.merged.module/javafx.beans.property.ObjectProperty.setValue(Unknown Source)
       at org.jabref/org.jabref.gui.StateManager.setSearchQuery(Unknown Source)
       at org.jabref/org.jabref.gui.search.GlobalSearchBar.performSearch(Unknown Source)
       at org.jabref/org.jabref.gui.search.GlobalSearchBar.lambda$new$8(Unknown Source)
       at org.jabref/org.jabref.gui.util.BindingsHelper$BidirectionalBinding.updateLocked(Unknown Source)
       at org.jabref/org.jabref.gui.util.BindingsHelper$BidirectionalBinding.changedB(Unknown Source)
       at org.jabref.merged.module/com.sun.javafx.binding.ExpressionHelper$Generic.fireValueChangedEvent(Unknown Source)
       at org.jabref.merged.module/com.sun.javafx.binding.ExpressionHelper.fireValueChangedEvent(Unknown Source)
       at org.jabref.merged.module/javafx.scene.control.TextInputControl$TextProperty.fireValueChangedEvent(Unknown Source)
       at org.jabref.merged.module/javafx.scene.control.TextInputControl$TextProperty.markInvalid(Unknown Source)
       at org.jabref.merged.module/javafx.scene.control.TextInputControl$TextProperty.controlContentHasChanged(Unknown Source)
       at org.jabref.merged.module/javafx.scene.control.TextInputControl.lambda$new$0(Unknown Source)
       at org.jabref.merged.module/com.sun.javafx.binding.ExpressionHelper$SingleInvalidation.fireValueChangedEvent(Unknown Source)
       at org.jabref.merged.module/com.sun.javafx.binding.ExpressionHelper.fireValueChangedEvent(Unknown Source)
       at org.jabref.merged.module/javafx.scene.control.TextField$TextFieldContent.insert(Unknown Source)
       at org.jabref.merged.module/javafx.scene.control.TextInputControl.replaceText(Unknown Source)
       at org.jabref.merged.module/javafx.scene.control.TextInputControl.updateContent(Unknown Source)
       at org.jabref.merged.module/javafx.scene.control.TextInputControl.replaceText(Unknown Source)
       at org.jabref.merged.module/javafx.scene.control.TextInputControl.replaceText(Unknown Source)
       at org.jabref.merged.module/javafx.scene.control.TextInputControl.replaceSelection(Unknown Source)
       at org.jabref.merged.module/javafx.scene.control.TextInputControl.paste(Unknown Source)
       at org.jabref.merged.module/com.sun.javafx.scene.control.behavior.TextInputControlBehavior.paste(Unknown Source)
       at org.jabref.merged.module/com.sun.javafx.scene.control.behavior.TextInputControlBehavior.lambda$new$20(Unknown Source)
       at org.jabref.merged.module/com.sun.javafx.scene.control.behavior.TextInputControlBehavior.lambda$keyMapping$62(Unknown Source)
       at org.jabref.merged.module/com.sun.javafx.scene.control.inputmap.InputMap.handle(Unknown Source)
       at org.jabref.merged.module/com.sun.javafx.event.CompositeEventHandler$NormalEventHandlerRecord.handleBubblingEvent(Unknown Source)
       at org.jabref.merged.module/com.sun.javafx.event.CompositeEventHandler.dispatchBubblingEvent(Unknown Source)
       at org.jabref.merged.module/com.sun.javafx.event.EventHandlerManager.dispatchBubblingEvent(Unknown Source)
       at org.jabref.merged.module/com.sun.javafx.event.EventHandlerManager.dispatchBubblingEvent(Unknown Source)
       at org.jabref.merged.module/com.sun.javafx.event.CompositeEventDispatcher.dispatchBubblingEvent(Unknown Source)
       at org.jabref.merged.module/com.sun.javafx.event.BasicEventDispatcher.dispatchEvent(Unknown Source)
       at org.jabref.merged.module/com.sun.javafx.event.EventDispatchChainImpl.dispatchEvent(Unknown Source)
       at org.jabref.merged.module/com.sun.javafx.event.BasicEventDispatcher.dispatchEvent(Unknown Source)
       at org.jabref.merged.module/com.sun.javafx.event.EventDispatchChainImpl.dispatchEvent(Unknown Source)
       at org.jabref.merged.module/com.sun.javafx.event.BasicEventDispatcher.dispatchEvent(Unknown Source)
       at org.jabref.merged.module/com.sun.javafx.event.EventDispatchChainImpl.dispatchEvent(Unknown Source)
       at org.jabref.merged.module/com.sun.javafx.event.BasicEventDispatcher.dispatchEvent(Unknown Source)
       at org.jabref.merged.module/com.sun.javafx.event.EventDispatchChainImpl.dispatchEvent(Unknown Source)
       at org.jabref.merged.module/com.sun.javafx.event.BasicEventDispatcher.dispatchEvent(Unknown Source)
       at org.jabref.merged.module/com.sun.javafx.event.EventDispatchChainImpl.dispatchEvent(Unknown Source)
       at org.jabref.merged.module/com.sun.javafx.event.EventUtil.fireEventImpl(Unknown Source)
       at org.jabref.merged.module/com.sun.javafx.event.EventUtil.fireEvent(Unknown Source)
       at org.jabref.merged.module/javafx.event.Event.fireEvent(Unknown Source)
       at org.jabref.merged.module/javafx.scene.Scene$KeyHandler.process(Unknown Source)
       at org.jabref.merged.module/javafx.scene.Scene.processKeyEvent(Unknown Source)
       at org.jabref.merged.module/javafx.scene.Scene$ScenePeerListener.keyEvent(Unknown Source)
       at org.jabref.merged.module/com.sun.javafx.tk.quantum.GlassViewEventHandler$KeyEventNotification.run(Unknown Source)
       at org.jabref.merged.module/com.sun.javafx.tk.quantum.GlassViewEventHandler$KeyEventNotification.run(Unknown Source)
       at java.base/java.security.AccessController.doPrivileged(Unknown Source)
       at org.jabref.merged.module/com.sun.javafx.tk.quantum.GlassViewEventHandler.lambda$handleKeyEvent$1(Unknown Source)
       at org.jabref.merged.module/com.sun.javafx.tk.quantum.QuantumToolkit.runWithoutRenderLock(Unknown Source)
       at org.jabref.merged.module/com.sun.javafx.tk.quantum.GlassViewEventHandler.handleKeyEvent(Unknown Source)
       at org.jabref.merged.module/com.sun.glass.ui.View.handleKeyEvent(Unknown Source)
       at org.jabref.merged.module/com.sun.glass.ui.View.notifyKey(Unknown Source)
       at org.jabref.merged.module/com.sun.glass.ui.win.WinApplication._runLoop(Native Method)
       at org.jabref.merged.module/com.sun.glass.ui.win.WinApplication.lambda$runLoop$3(Unknown Source)
       at java.base/java.lang.Thread.run(Unknown Source)
ERROR FallbackExceptionHandler Uncaught exception occurred in Thread[JavaFX Application Thread,5,main]
netscape.javascript.JSException: SyntaxError: Invalid regular expression: missing )
       at org.jabref.merged.module/com.sun.webkit.dom.JSObject.fwkMakeException(Unknown Source)
       at org.jabref.merged.module/com.sun.webkit.WebPage.twkExecuteScript(Native Method)
       at org.jabref.merged.module/com.sun.webkit.WebPage.executeScript(Unknown Source)
       at org.jabref.merged.module/javafx.scene.web.WebEngine.executeScript(Unknown Source)
       at org.jabref/org.jabref.gui.preview.PreviewViewer.highlightSearchPattern(Unknown Source)
       at org.jabref/org.jabref.gui.preview.PreviewViewer.lambda$new$0(Unknown Source)
       at org.jabref.merged.module/com.sun.javafx.binding.ExpressionHelper$Generic.fireValueChangedEvent(Unknown Source)
       at org.jabref.merged.module/com.sun.javafx.binding.ExpressionHelper.fireValueChangedEvent(Unknown Source)
       at org.jabref.merged.module/javafx.beans.property.ObjectPropertyBase.fireValueChangedEvent(Unknown Source)
       at org.jabref.merged.module/javafx.beans.property.ObjectPropertyBase.markInvalid(Unknown Source)
       at org.jabref.merged.module/javafx.beans.property.ObjectPropertyBase.set(Unknown Source)
       at org.jabref.merged.module/javafx.beans.property.ObjectProperty.setValue(Unknown Source)
       at org.jabref/org.jabref.gui.StateManager.setSearchQuery(Unknown Source)
       at org.jabref/org.jabref.gui.search.GlobalSearchBar.performSearch(Unknown Source)
       at org.jabref.merged.module/org.reactfx.util.FxTimer.lambda$restart$0(Unknown Source)
       at org.jabref.merged.module/com.sun.scenario.animation.shared.TimelineClipCore.visitKeyFrame(Unknown Source)
       at org.jabref.merged.module/com.sun.scenario.animation.shared.TimelineClipCore.playTo(Unknown Source)
       at org.jabref.merged.module/javafx.animation.Timeline.doPlayTo(Unknown Source)
       at org.jabref.merged.module/javafx.animation.AnimationAccessorImpl.playTo(Unknown Source)
       at org.jabref.merged.module/com.sun.scenario.animation.shared.SingleLoopClipEnvelope.timePulse(Unknown Source)
       at org.jabref.merged.module/javafx.animation.Animation.doTimePulse(Unknown Source)
       at org.jabref.merged.module/javafx.animation.Animation$1.lambda$timePulse$0(Unknown Source)
       at java.base/java.security.AccessController.doPrivileged(Unknown Source)
       at org.jabref.merged.module/javafx.animation.Animation$1.timePulse(Unknown Source)
       at org.jabref.merged.module/com.sun.scenario.animation.AbstractMasterTimer.timePulseImpl(Unknown Source)
       at org.jabref.merged.module/com.sun.scenario.animation.AbstractMasterTimer$MainLoop.run(Unknown Source)
       at org.jabref.merged.module/com.sun.javafx.tk.quantum.QuantumToolkit.pulse(Unknown Source)
       at org.jabref.merged.module/com.sun.javafx.tk.quantum.QuantumToolkit.pulse(Unknown Source)
       at org.jabref.merged.module/com.sun.javafx.tk.quantum.QuantumToolkit.pulseFromQueue(Unknown Source)
       at org.jabref.merged.module/com.sun.javafx.tk.quantum.QuantumToolkit.lambda$runToolkit$11(Unknown Source)
       at org.jabref.merged.module/com.sun.glass.ui.InvokeLaterDispatcher$Future.run(Unknown Source)
       at org.jabref.merged.module/com.sun.glass.ui.win.WinApplication._runLoop(Native Method)
       at org.jabref.merged.module/com.sun.glass.ui.win.WinApplication.lambda$runLoop$3(Unknown Source)
       at java.base/java.lang.Thread.run(Unknown Source)
  ```
</details>

### PR DESCRIPTION
**Pull request by dawidm**
fixes #5892

The error when searching for "DOI 10.1210/endrev/bnz006" (or any phrases containing special javascript regular expression characters) is caused by not escaping special characters before passing search  pattern to javascript script used for highlighting words in preview view.

- [ ] Change in CHANGELOG.md described (if applicable)
- [x] Tests created for changes (if applicable)
- [x] Manually tested changed features in running JabRef (always required)
- [ ] Screenshots added in PR description (for bigger UI changes)
- [x] [Checked documentation](https://docs.jabref.org/): Is the information available and up to date? If not: Issue created at <https://github.com/JabRef/user-documentation/issues>.

### CODE REVIEW

File: src/main/java/org/jabref/logic/search/SearchQuery.java [hunk 1]
```diff
@@ -124,6 +127,18 @@ public boolean isRegularExpression() {
...
+
+    // Returns a regular expression pattern in the form (w1)|(w2)| ... wi are escaped for javascript if no regular expression search is enabled
+    public Optional<Pattern> getJsPatternForWords() {
+        return joinWordsToPattern(true);
+    }
+
+    /* Returns a regular expression pattern in the form (w1)|(w2)| ... wi are escaped if no regular expression search is enabled
+     * @param escapeSpecialCharsForJS whether to escape characters in wi for javascript regexp (escaping all special characters) or for java (using \Q and \E)
+     */
+    private Optional<Pattern> joinWordsToPattern(boolean escapeSpecialCharsForJS) {
```
Comment (Siedlerchr) [inline src/main/java/org/jabref/logic/search/SearchQuery.java]: Better is to use an enum here, eg.. EscapMode, with java and javascript as values:
https://www.teamten.com/lawrence/programming/prefer-enums-over-booleans.html

File: src/main/java/org/jabref/logic/search/SearchQuery.java [hunk 2]
```diff
@@ -18,6 +18,9 @@
 
 public class SearchQuery implements SearchMatcher {
 
+    // regexp pattern for escaping special characters in javascript regex
+    public static final String JAVASCRIPT_ESCAPED_CHARS_PATTERN = "[\\.\\*\\+\\?\\^\\$\\{\\}\\(\\)\\|\\[\\]\\\\/]";
```
Comment (Siedlerchr) [inline src/main/java/org/jabref/logic/search/SearchQuery.java]: You can use Pattern.compile to gain some performace improvements here

File: src/main/java/org/jabref/logic/search/SearchQuery.java [hunk 3]
```diff
@@ -133,7 +148,7 @@ public boolean isRegularExpression() {
         // compile the words to a regular expression in the form (w1)|(w2)|(w3)
         StringJoiner joiner = new StringJoiner(")|(", "(", ")");
         for (String word : words) {
-            joiner.add(regularExpression ? word : Pattern.quote(word));
+            joiner.add(regularExpression ? word : (escapeSpecialCharsForJS ? word.replaceAll(JAVASCRIPT_ESCAPED_CHARS_PATTERN, "\\\\$0") : Pattern.quote(word)));
```
Comment (Siedlerchr) [inline src/main/java/org/jabref/logic/search/SearchQuery.java]: Please extract this to a regular if-else, it's easier to understand on the first look than this chained conditions

File: src/main/java/org/jabref/logic/search/SearchQuery.java [hunk 4]
```diff
@@ -133,7 +164,21 @@ public boolean isRegularExpression() {
         // compile the words to a regular expression in the form (w1)|(w2)|(w3)
         StringJoiner joiner = new StringJoiner(")|(", "(", ")");
         for (String word : words) {
-            joiner.add(regularExpression ? word : Pattern.quote(word));
+            if (regularExpression) {
+                joiner.add(word);
+            }
+            else {
```
Comment (calixtus) [inline src/main/java/org/jabref/logic/search/SearchQuery.java]: Isn't checkstyle complaining about putting this in two lines? Has no effect, but looks somewhat odd...

File: src/main/java/org/jabref/logic/search/SearchQuery.java [hunk 5]
```diff
@@ -124,6 +143,18 @@ public boolean isRegularExpression() {
 
     // Returns a regular expression pattern in the form (w1)|(w2)| ... wi are escaped if no regular expression search is enabled
     public Optional<Pattern> getPatternForWords() {
+        return joinWordsToPattern(EscapeMode.JAVA);
+    }
+
+    // Returns a regular expression pattern in the form (w1)|(w2)| ... wi are escaped for javascript if no regular expression search is enabled
+    public Optional<Pattern> getJsPatternForWords() {
```
Comment (calixtus) [inline src/main/java/org/jabref/logic/search/SearchQuery.java]: IMHO, although "Js" is quite obvious, I would reword that to getJavaScriptPatternForWords, since you are creating a new Method just for the sole purpose to make its use more obvious in it's name. So why not go all the way?

File: src/main/java/org/jabref/logic/search/SearchQuery.java [hunk 6]
```diff
@@ -18,6 +18,25 @@
 
 public class SearchQuery implements SearchMatcher {
 
+    /**
+     * Regex pattern for escaping special characters in javascript regular expressions
+     */
+    public static final Pattern JAVASCRIPT_ESCAPED_CHARS_PATTERN = Pattern.compile("[\\.\\*\\+\\?\\^\\$\\{\\}\\(\\)\\|\\[\\]\\\\/]");
+
+    /**
+     * Metod for escaping special characters in regular expressions
```
Comment (stefan-kolb) [inline src/main/java/org/jabref/logic/search/SearchQuery.java]: Method

File: src/main/java/org/jabref/logic/search/SearchQuery.java (hunk 4 above)
Comment (dawidm) [reply src/main/java/org/jabref/logic/search/SearchQuery.java]: Yeah, it's a mistake, I'll change it. Also looks odd to me. Checkstyle wasn't complaining.

**Comment by dawidm** [comment]
> A test would be great...

Ok, I will write a test.

File: src/test/java/org/jabref/logic/search/SearchQueryTest.java [hunk 7]
```diff
@@ -203,4 +203,38 @@ public void testGetPattern() {
...
+        String pattern = "(\\Q<([{\\^-=$!|]})?*+.>\\E)|(\\Qword1\\E)|(\\Qword2.\\E)";
+        assertEquals(Optional.of(pattern), textQueryWithSpecialChars.getPatternForWords().map(Pattern::toString));
+    }
+
+    @Test
+    public void testEscapingInJavascriptPattern() {
+        //first word contain all javascript special regex characters that should be escaped individually in text based search
+        String queryText = "([{\\\\^$|]})?*+./ word1 word2.";
+        SearchQuery textQueryWithSpecialChars = new SearchQuery(queryText, false, false);
+        String pattern = "(\\(\\[\\{\\\\\\^\\$\\|\\]\\}\\)\\?\\*\\+\\.\\/)|(word1)|(word2\\.)";
```
Comment (dawidm) [inline src/test/java/org/jabref/logic/search/SearchQueryTest.java]: I've created tests but I'm not sure about readbility of this approach.

Comment (Siedlerchr) [reply src/test/java/org/jabref/logic/search/SearchQueryTest.java]: I know this looks totally crazy with those escaping, but it's fine. It's

**Comment by calixtus** [comment]
I think this is really great as well! Thank you!
